### PR TITLE
En 678 api pagination sorting filtering business layer

### PR DESCRIPTION
**Pull request by stefanopuddu**
Update the *searcherDAO classes

### CODE REVIEW

**Comment by entando-service** [comment]
Something went wrong. Investigate!

File: engine/src/main/java/com/agiletec/aps/system/common/AbstractSearcherDAO.java [hunk 1]
```diff
@@ -32,464 +32,469 @@
...
+     * @param stat The statement where the filters are applied.
+     * @return The last used index.
+     * @throws SQLException In case of error.
+     *
+     */
+    protected int addObjectSearchStatementBlock(FieldSearchFilter filter, int index, PreparedStatement stat) throws SQLException {
+        return addSearchStatementBlock(filter, index, stat);
+    }
+
+    protected int addObjectSearchStatementBlockForCount(FieldSearchFilter filter, int index, PreparedStatement stat) throws SQLException {
```
Comment (joewhite101) [inline engine/src/main/java/com/agiletec/aps/system/common/AbstractSearcherDAO.java]: This wrapper around addSearchStatementBlock isn't needed. Remove

File: engine/src/main/java/com/agiletec/aps/system/common/AbstractSearcherDAO.java [hunk 2]
```diff
@@ -32,464 +32,469 @@
...
+    /**
+     * Add to the statement a filter on a attribute.
+     * @param filter The filter on the attribute to apply in the statement.
+     * @param index The last index used to associate the elements to the statement.
+     * @param stat The statement where the filters are applied.
+     * @return The last used index.
+     * @throws SQLException In case of error.
+     *
+     */
+    protected int addObjectSearchStatementBlock(FieldSearchFilter filter, int index, PreparedStatement stat) throws SQLException {
```
Comment (joewhite101) [inline engine/src/main/java/com/agiletec/aps/system/common/AbstractSearcherDAO.java]: Is this needed? Looks like a pass through to addSearchStatementBlock

File: engine/src/main/java/com/agiletec/aps/system/common/AbstractSearcherDAO.java [hunk 3]
```diff
@@ -32,464 +32,469 @@
...
+        for (int i = 0; i < filters.length; i++) {
+            FieldSearchFilter filter = filters[i];
+            if (filter.getKey() != null) {
+                hasAppendWhereClause = this.addMetadataFieldFilterQueryBlock(filter, query, hasAppendWhereClause);
+            }
+        }
+        return hasAppendWhereClause;
+    }
+
+    protected boolean addMetadataFieldFilterQueryBlockForCount(FieldSearchFilter filter, StringBuffer query, boolean hasAppendWhereClause) {
```
Comment (joewhite101) [inline engine/src/main/java/com/agiletec/aps/system/common/AbstractSearcherDAO.java]: This method is the same  as addMetadataFieldFilterQueryBlock. Remove

File: engine/src/main/java/com/agiletec/aps/system/common/AbstractSearcherDAO.java [hunk 4]
```diff
@@ -32,464 +32,469 @@
...
+        } else {
+            stat.setObject(index, object);
+        }
+    }
+
+    protected String createQueryString(FieldSearchFilter[] filters, boolean isCount, boolean selectAll) {
+        StringBuffer query = this.createBaseQueryBlock(filters, isCount, selectAll);
+
+        boolean hasAppendWhereClause = false;
+        if (isCount) {
```
Comment (joewhite101) [inline engine/src/main/java/com/agiletec/aps/system/common/AbstractSearcherDAO.java]: This if statement isn't needed. The call to 
appendMetadataFieldFilterQueryBlocksForCount
and
appendMetadataFieldFilterQueryBlocksFor

End up making the same filter calls

File: engine/src/main/java/com/agiletec/aps/system/common/AbstractSearcherDAO.java [hunk 5]
```diff
@@ -32,464 +32,469 @@
...
+                        query.append(", ").append(masterTableName).append(".").append(this.getTableFieldName(filters[i].getKey()));
+                    }
+                }
+            }
+        }
+        query.append(" FROM ").append(masterTableName).append(" ");
+        return query;
+    }
+
+    //XXX REMOVE VARS per prepared statement
```
Comment (joewhite101) [inline engine/src/main/java/com/agiletec/aps/system/common/AbstractSearcherDAO.java]: Commend needed or did this get done?

File: engine/src/main/java/com/agiletec/plugins/jacms/aps/system/services/content/PublicContentSearcherDAO.java [hunk 6]
```diff
@@ -74,7 +73,7 @@
 			conn = this.getConnection();
 			stat = this.buildStatement(filters, categories, orClauseCategoryFilter, groupCodes, false, conn);
 			result = stat.executeQuery();
-			this.flowResult(contentsId, filters, result);
+            //this.flowResult(contentsId, filters, result);
```
Comment (joewhite101) [inline engine/src/main/java/com/agiletec/plugins/jacms/aps/system/services/content/PublicContentSearcherDAO.java]: Remove

File: engine/src/main/java/com/agiletec/plugins/jacms/aps/system/services/content/WorkContentSearcherDAO.java [hunk 7]
```diff
@@ -56,7 +55,7 @@
 			conn = this.getConnection();
 			stat = this.buildStatement(filters, categories, orClauseCategoryFilter, userGroupCodes, false, conn);
 			result = stat.executeQuery();
-			this.flowResult(contentsId, filters, result);
+            //this.flowResult(contentsId, filters, result);
```
Comment (joewhite101) [inline engine/src/main/java/com/agiletec/plugins/jacms/aps/system/services/content/WorkContentSearcherDAO.java]: Remove

File: engine/src/main/java/com/agiletec/plugins/jacms/aps/system/services/resource/ResourceDAO.java [hunk 8]
```diff
@@ -244,7 +243,7 @@ protected void executeDeleteResource(String resourceId, Connection conn) throws
 			conn = this.getConnection();
 			stat = this.buildStatement(filters, categoryCode, groupCodes, conn);
 			res = stat.executeQuery();
-			this.flowResult(resources, filters, res);
+            //this.flowResult(resources, filters, res);
```
Comment (joewhite101) [inline engine/src/main/java/com/agiletec/plugins/jacms/aps/system/services/resource/ResourceDAO.java]: Remove

File: engine/src/main/java/org/entando/entando/aps/system/services/dataobject/DataObjectSearcherDAO.java [hunk 9]
```diff
@@ -74,7 +73,15 @@
             conn = this.getConnection();
             stat = this.buildStatement(filters, categories, orClauseCategoryFilter, groupCodes, false, conn);
             result = stat.executeQuery();
-            this.flowResult(contentsId, filters, result);
+            //this.flowResult(contentsId, filters, result);
```
Comment (joewhite101) [inline engine/src/main/java/org/entando/entando/aps/system/services/dataobject/DataObjectSearcherDAO.java]: Remove

File: engine/src/main/java/com/agiletec/aps/system/common/AbstractSearcherDAO.java [hunk 10]
```diff
@@ -32,464 +32,469 @@
...
+        } catch (Throwable t) {
+            throw new RuntimeException("error building limit query", t);
+        }
+    }
+
+    protected boolean appendMetadataFieldFilterQueryBlocksForCount(FieldSearchFilter[] filters, StringBuffer query, boolean hasAppendWhereClause) {
+        if (filters == null) {
+            return hasAppendWhereClause;
+        }
+        for (int i = 0; i < filters.length; i++) {
```
Comment (joewhite101) [inline engine/src/main/java/com/agiletec/aps/system/common/AbstractSearcherDAO.java]: Use new style loops.

File: engine/src/main/java/com/agiletec/aps/system/common/AbstractSearcherDAO.java [hunk 11]
```diff
@@ -32,464 +32,469 @@
...
+            }
+        }
+        return hasAppendWhereClause;
+    }
+
+    protected boolean appendMetadataFieldFilterQueryBlocks(FieldSearchFilter[] filters, StringBuffer query, boolean hasAppendWhereClause) {
+        if (filters == null) {
+            return hasAppendWhereClause;
+        }
+        for (int i = 0; i < filters.length; i++) {
```
Comment (joewhite101) [inline engine/src/main/java/com/agiletec/aps/system/common/AbstractSearcherDAO.java]: Update loop structure.

File: engine/src/main/java/com/agiletec/aps/system/common/AbstractSearcherDAO.java [hunk 12]
```diff
@@ -32,464 +32,469 @@
...
+            }
+        }
+        return hasAppendWhereClause;
+    }
+
+    protected boolean appendOrderQueryBlocks(FieldSearchFilter[] filters, StringBuffer query, boolean ordered) {
+        if (filters == null) {
+            return ordered;
+        }
+        for (int i = 0; i < filters.length; i++) {
```
Comment (joewhite101) [inline engine/src/main/java/com/agiletec/aps/system/common/AbstractSearcherDAO.java]: Loop structure

File: engine/src/main/java/com/agiletec/aps/system/common/AbstractSearcherDAO.java [hunk 13]
```diff
@@ -32,464 +32,469 @@
...
+
+    public void setLikeClause(String likeClause) {
+        this._likeClause = likeClause;
+    }
+
+    protected boolean hasLikeFilters(FieldSearchFilter[] filters) {
+        if (null == filters || filters.length == 0) {
+            return false;
+        }
+        for (int i = 0; i < filters.length; i++) {
```
Comment (joewhite101) [inline engine/src/main/java/com/agiletec/aps/system/common/AbstractSearcherDAO.java]: Loop structure

File: engine/src/main/java/com/agiletec/aps/system/common/QueryLimitResolver.java [hunk 14]
```diff
@@ -0,0 +1,72 @@
...
+
+    public static String createLimitBlock(Integer offset, Integer limit, DataSource dataSource) throws NoSuchMethodException, SecurityException, IllegalAccessException, IllegalArgumentException, InvocationTargetException {
+        String limitBlock = null;
+        String driverClassName = extractDriverClassName(dataSource);
+        logger.trace("detected driver: {}", driverClassName);
+        if (driverClassName.equalsIgnoreCase(JDBC_DRIVER_DERBY_EMBEDDED)) {
+            limitBlock = String.format(" OFFSET %d ROWS FETCH NEXT %d ROWS ONLY ", new Object[]{offset, limit});
+        } else if (driverClassName.equalsIgnoreCase(JDBC_DRIVER_POSTGRES)) {
+            limitBlock = String.format(" OFFSET %d ROWS FETCH NEXT %d ROWS ONLY ", new Object[]{offset, limit});
+        } else if (driverClassName.equalsIgnoreCase("TODO")) {
```
Comment (joewhite101) [inline engine/src/main/java/com/agiletec/aps/system/common/QueryLimitResolver.java]: TODO? Haven't encountered this driver before ... ;)

File: engine/src/main/java/com/agiletec/aps/system/common/QueryLimitResolver.java [hunk 15]
```diff
@@ -0,0 +1,72 @@
...
+            throw new UnsupportedOperationException(driverClassName + " not implemented!");
+        }  else {
+            throw new UnsupportedOperationException(driverClassName + " not implemented!");
```
Comment (joewhite101) [inline engine/src/main/java/com/agiletec/aps/system/common/QueryLimitResolver.java]: Log here.

File: engine/src/main/java/com/agiletec/aps/system/common/AbstractSearcherDAO.java [hunk 16]
```diff
@@ -32,464 +32,469 @@
...
+     * is selected.
+     * @return The base block of the query.
+     */
+    protected StringBuffer createBaseQueryBlock(FieldSearchFilter[] filters, boolean selectAll) {
+        return this.createBaseQueryBlock(filters, false, selectAll);
+    }
+
+    protected StringBuffer createBaseQueryBlock(FieldSearchFilter[] filters, boolean isCount, boolean selectAll) {
+        StringBuffer query = null;
+        if (isCount) {
```
Comment (joewhite101) [inline engine/src/main/java/com/agiletec/aps/system/common/AbstractSearcherDAO.java]: If statement not needed. Calls are the same.

If the calls are the same is the method needed.

**Comment by entando-service** [comment]
Something went wrong. Investigate!

**Comment by entando-service** [comment]
Something went wrong. Investigate!

File: engine/src/main/java/com/agiletec/aps/system/common/QueryLimitResolver.java (hunk 14 above)
Comment (stefanopuddu) [reply engine/src/main/java/com/agiletec/aps/system/common/QueryLimitResolver.java]: this is a placeholder for mysql :)

**Comment by entando-service** [comment]
Something went wrong. Investigate!

**Comment by entando-service** [comment]
Something went wrong. Investigate!

**Comment by entando-service** [comment]
Something went wrong. Investigate!

**Comment by entando-service** [comment]
Something went wrong. Investigate!

**Comment by entando-service** [comment]
All is well

**Pull request merged**